### PR TITLE
Fix empty input issue of convolution for channels last memory format

### PR DESCRIPTION
**Pull request by CaoE**
Fixes empty input convolution issue : when input is empty e.g. shape of (0, 3, 3, 4) and weight is channels last format, at::_unsafe_view will raise "view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead."

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by yanbing-j** [comment]
Need merge access for test_nn.py.

**Comment by CaoE** [comment]
Hi @frank-wei , could you please review this PR, thank you.

**Review by malfet — CHANGES_REQUESTED** [review]
I don't know how frequent operations on empty tensors, but it feels like requesting a contiguous copy of a weighs tensor just to compute shape of an output tensor is a waste of resources. 
Can we just detect that memory type is channels_last and call shuffle?

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -1454,7 +1454,8 @@ at::Tensor _convolution(
       break;
     case ConvBackend::Empty:
     {
-      auto weight_view = at::_unsafe_view(weight, -1);
+      auto weight_view = weight.is_contiguous() ? at::_unsafe_view(weight, -1) :
```
Comment (malfet) [inline aten/src/ATen/native/Convolution.cpp:1457]: It would be good to add some sort of comment, why this is needed

Comment (CaoE) [reply aten/src/ATen/native/Convolution.cpp]: Add comments

**Comment by CaoE** [comment]
> I don't know how frequent operations on empty tensors, but it feels like requesting a contiguous copy of a weighs tensor just to compute shape of an output tensor is a waste of resources. Can we just detect that memory type is channels_last and call shuffle?

Yes, it is a waste of resources. Using permute instead clone will be better for channels_last contiguous cases.

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -1454,7 +1453,19 @@ at::Tensor _convolution(
...
+      // Use permute and clone to avoid at::_unsafe_view(weight, -1) failure for non-contiguous cases where
+      // view size is not compatible with input tensor's size and stride.
+      if(weight.is_contiguous()) {
+        weight_view = at::_unsafe_view(weight, -1);
+      } else if (weight.is_contiguous(at::MemoryFormat::ChannelsLast)) {
+        weight_view = at::_unsafe_view(at::permute(weight, {0, 2, 3, 1}), -1);
+      } else if (weight.is_contiguous(at::MemoryFormat::ChannelsLast3d)) {
+        weight_view = at::_unsafe_view(at::permute(weight, {0, 2, 3, 4, 1}), -1);
+      } else {
+        weight_view = at::_unsafe_view(weight.clone(at::MemoryFormat::Contiguous), -1);
```
Comment (jgong5) [inline aten/src/ATen/native/Convolution.cpp:1466]: consider add a test to cover this scenario?

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/Convolution.cpp:1472]: Add the noncontiguous case.

File: test/test_nn.py [hunk 3]
```diff
@@ -740,6 +740,34 @@ def test_no_grad(self):
                 self.assertFalse(output2.requires_grad)
                 self.assertRaises(RuntimeError, lambda: output2.backward(torch.ones(1, 5, 10, 10)))
```
Comment (mingfeima) [inline test/test_nn.py:742]: Need to add `device` and `dtype` for the test case, otherwise it will only test on CPU.

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: test/test_nn.py [hunk 4]
```diff
@@ -740,6 +740,34 @@ def test_no_grad(self):
...
 
+    def test_conv_empty_input(self):
```
Comment (mingfeima) [inline test/test_nn.py:743]: Need to add `device` and `dtype` for the test case, otherwise it will only test on CPU.

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: test/test_nn.py (hunk 4 above)
Comment (CaoE) [reply test/test_nn.py]: Added `device` and `dtype`

**Comment by CaoE** [comment]
Hi @malfet, could you please review this PR ? Thank you.

**Review by malfet — APPROVED** [review]
